File: src/utils/role.ts
// 角色常量定义
export const ROLE_USER = 1      // 普通用户
export const ROLE_DEVELOPER = 2 // 开发者
export const ROLE_ADMIN = 4     // 管理员
export const ROLE_AUDITOR = 8   // 审核员
export const ROLE_OPERATOR = 16 // 运营

// 角色信息接口
export interface RoleInfo {
  value: number
  key: string
  name: string
  description: string
}

// 所有角色列表
export const ALL_ROLES: RoleInfo[] = [
  { value: ROLE_USER, key: 'user', name: '普通用户', description: '浏览、下载、安装微应用' },
  { value: ROLE_DEVELOPER, key: 'developer', name: '开发者', description: '创建和管理微应用' },
  { value: ROLE_ADMIN, key: 'admin', name: '管理员', description: '系统管理、用户管理' },
  { value: ROLE_AUDITOR, key: 'auditor', name: '审核员', description: '审核微应用版本' },
  { value: ROLE_OPERATOR, key: 'operator', name: '运营', description: '数据统计、推广管理' },
]

/**
 * 判断用户是否拥有某个角色
 */
export function hasRole(userRole: number, role: number): boolean {
  return (userRole & role) !== 0
}

/**
 * 判断用户是否是管理员
 */
export function hasAdmin(userRole: number): boolean {
  return hasRole(userRole, ROLE_ADMIN)
}

/**
 * 添加角色
 */
export function addRole(userRole: number, role: number): number {
  return userRole | role
}

/**
 * 移除角色
 */
export function removeRole(userRole: number, role: number): number {
  return userRole & ~role
}

/**
 * 获取用户的所有角色列表
 */
export function getRoleList(userRole: number): RoleInfo[] {
  return ALL_ROLES.filter(role => hasRole(userRole, role.value))
}

/**
 * 获取用户所有角色名称列表
 */
export function getRoleNames(userRole: number): string[] {
  return ALL_ROLES.filter(role => hasRole(userRole, role.value)).map(role => role.name)
}

/**
 * 根据角色值获取角色标签颜色
 */
export function getRoleTagType(role: number): 'default' | 'error' | 'primary' | 'success' | 'warning' {
  switch (role) {
    case ROLE_ADMIN:
      return 'error'
    case ROLE_DEVELOPER:
      return 'primary'
    case ROLE_AUDITOR:
      return 'success'
    case ROLE_OPERATOR:
      return 'warning'
    default:
      return 'default'
  }
}

/**
 * 切换角色
 */
export function toggleRole(userRole: number, role: number): number {
  if (hasRole(userRole, role)) {
    return removeRole(userRole, role)
  }
  return addRole(userRole, role)
}

/**
 * 计算选中的角色列表（用于复选框组）
 */
export function getSelectedRoles(userRole: number): number[] {
  return ALL_ROLES.filter(role => hasRole(userRole, role.value)).map(role => role.value)
}

/**
 * 从选中列表计算角色值
 */
export function calculateRoleFromSelected(selectedRoles: number[]): number {
  return selectedRoles.reduce((acc, role) => acc | role, 0)
}

// ==================== 菜单权限 ====================

/**
 * 后台管理菜单项配置
 * - key: 路由名称（叶子节点必填）
 * - label: 菜单显示文本
 * - children: 子菜单
 * - roles: 需要的角色（多个角色用位运算组合），父级必填，子项可选（继承父级）
 */

// 叶子菜单项（无children）
export interface LeafMenuItem {
  key: string
  label: string
  roles?: number
}

// 分组菜单项（有children）
export interface GroupMenuItem {
  key: string
  label: string
  children: AdminMenuItem[]
  roles: number
}

// 联合类型
export type AdminMenuItem = LeafMenuItem | GroupMenuItem

/**
 * 递归过滤菜单：检查用户角色权限，子项未设置roles时继承父项
 */
export function filterAdminMenuByRole(
  userRole: number,
  menu: AdminMenuItem,
  parentRoles?: number
): AdminMenuItem | undefined {
  // 当前项的角色：自身roles或继承父项
  const currentRoles = menu.roles ?? parentRoles

  // 有children的是分组菜单，递归处理子项
  if ('children' in menu && menu.children && menu.children.length > 0) {
    const filteredChildren = menu.children
      .map(child => filterAdminMenuByRole(userRole, child, currentRoles))
      .filter(Boolean) as AdminMenuItem[]

    // 有可见子菜单时才返回
    if (filteredChildren.length > 0) {
      return { ...menu, children: filteredChildren }
    }
    return undefined
  }

  // 叶子菜单项，检查权限
  return hasRole(userRole, currentRoles!) ? menu : undefined
}

/**
 * 过滤并转换菜单为 naive-ui 格式
 */
export function filterAndConvertMenu(
  userRole: number,
  menuConfig: AdminMenuItem[]
): any[] {
  return menuConfig
    .map(item => filterAdminMenuByRole(userRole, item))
    .filter(Boolean)
}

File: service/models/role.go
package models

// 角色常量定义 - 使用位运算支持角色叠加
const (
	ROLE_USER      int = 1  // 普通用户   (二进制: 00001)
	ROLE_DEVELOPER int = 2  // 开发者     (二进制: 00010)
	ROLE_ADMIN     int = 4  // 管理员     (二进制: 00100)
	ROLE_AUDITOR   int = 8  // 审核员     (二进制: 01000)
	ROLE_OPERATOR  int = 16 // 运营       (二进制: 10000)
)

// 角色信息结构体
type RoleInfo struct {
	Value       int    `json:"value"`
	Key         string `json:"key"`
	Name        string `json:"name"`
	Description string `json:"description"`
}

// 所有角色列表
var AllRoles = []RoleInfo{
	{Value: ROLE_USER, Key: "user", Name: "普通用户", Description: "浏览、下载、安装微应用"},
	{Value: ROLE_DEVELOPER, Key: "developer", Name: "开发者", Description: "创建和管理微应用"},
	{Value: ROLE_ADMIN, Key: "admin", Name: "管理员", Description: "系统管理、用户管理"},
	{Value: ROLE_AUDITOR, Key: "auditor", Name: "审核员", Description: "审核微应用版本"},
	{Value: ROLE_OPERATOR, Key: "operator", Name: "运营", Description: "数据统计、推广管理"},
}

// HasRole 判断用户是否拥有某个角色
func HasRole(userRole int, role int) bool {
	return (userRole & role) != 0
}

// HasAdmin 判断用户是否是管理员
// 兼容旧系统：旧 Role=1 是管理员，新 Role=4 是管理员
func HasAdmin(userRole int) bool {
	// 新系统位运算判断
	if HasRole(userRole, ROLE_ADMIN) {
		return true
	}
	// 兼容旧系统：Role=1 是管理员
	if userRole == 1 {
		return true
	}
	return false
}

// AddRole 添加角色
func AddRole(userRole int, role int) int {
	return userRole | role
}

// RemoveRole 移除角色
func RemoveRole(userRole int, role int) int {
	return userRole &^ role
}

// GetRoleList 获取用户的所有角色列表
func GetRoleList(userRole int) []RoleInfo {
	roles := []RoleInfo{}
	for _, role := range AllRoles {
		if HasRole(userRole, role.Value) {
			roles = append(roles, role)
		}
	}
	return roles
}

// GetRoleNames 获取用户所有角色名称列表
func GetRoleNames(userRole int) []string {
	names := []string{}
	for _, role := range AllRoles {
		if HasRole(userRole, role.Value) {
			names = append(names, role.Name)
		}
	}
	return names
}

// MigrateOldRole 迁移旧角色值到新角色值
// 旧: Role=1(管理员), Role=2(普通用户)
// 新: Role=4(管理员), Role=1(普通用户)
func MigrateOldRole(oldRole int) int {
	switch oldRole {
	case 1: // 旧管理员
		return ROLE_ADMIN
	case 2: // 旧普通用户
		return ROLE_USER
	default:
		return oldRole
	}
}
